presto-main/src/main/java/com/facebook/presto/operator/GenericPageProcessor.java

```diff
@@ -36,16 +51,153 @@ public GenericPageProcessor(FilterFunction filterFunction, Iterable<? extends Pr
+
         for (; position < end && !pageBuilder.isFull(); position++) {
             if (filterFunction.filter(position, page.getBlocks())) {
                 pageBuilder.declarePosition();
                 for (int i = 0; i < projections.size(); i++) {
                     // todo: if the projection function increases the size of the data significantly, this could cause the servers to OOM
-                    projections.get(i).project(position, page.getBlocks(), pageBuilder.getBlockBuilder(i));
+                    projections.get(i).project(position, inputBlocks, pageBuilder.getBlockBuilder(i));
                 }
             }
         }
-
         return position;
     }
+
+    @Override
+    public Page processColumnar(ConnectorSession session, Page page, List<? extends Type> types)
+    {
+        int[] selectedPositions = filterPage(page);
+        if (selectedPositions.length == 0) {
+            return null;
+        }
+
+        if (projections.isEmpty()) {
+            return new Page(selectedPositions.length);
+        }
+
+        PageBuilder pageBuilder = new PageBuilder(types);
+        Block[] inputBlocks = page.getBlocks();
+
+        for (int projectionIndex = 0; projectionIndex < projections.size(); projectionIndex++) {
+            ProjectionFunction projection = projections.get(projectionIndex);
+            BlockBuilder blockBuilder = pageBuilder.getBlockBuilder(projectionIndex);
+            for (int position : selectedPositions) {
+                projection.project(position, inputBlocks, blockBuilder);
+            }
+        }
+        pageBuilder.declarePositions(selectedPositions.length);
+        return pageBuilder.build();
+    }
+
+    @Override
+    public Page processColumnarDictionary(ConnectorSession session, Page page, List<? extends Type> types)
+    {
+        Page inputPage = getNonLazyPage(page);
+        int[] selectedPositions = filterPage(inputPage);
+
+        if (selectedPositions.length == 0) {
+            return null;
+        }
+
+        if (projections.isEmpty()) {
+            return new Page(selectedPositions.length);
+        }
+
+        PageBuilder pageBuilder = new PageBuilder(types);
+        Block[] inputBlocks = page.getBlocks();
+        Block[] outputBlocks = new Block[projections.size()];
+
+        for (int projectionIndex = 0; projectionIndex < projections.size(); projectionIndex++) {
+            ProjectionFunction projection = projections.get(projectionIndex);
+
+            if (canDictionaryProcess(projection, inputPage)) {
+                Block outputDictionary = projectDictionary(projection, inputPage);
+                int[] outputIds = filterIds(projection, inputPage, selectedPositions);
+                outputBlocks[projectionIndex] = new DictionaryBlock(selectedPositions.length, outputDictionary, wrappedIntArray(outputIds));
+            }
+            else {
+                BlockBuilder blockBuilder = pageBuilder.getBlockBuilder(projectionIndex);
```

Comment:
maybe pull this out into a `processColumnar` method, and share with above method